docs/design/design.md

```diff
@@ -0,0 +1,116 @@
+* MicroShift does not bundle any OS user space! Bundling makes maintenance and security hard, breaks compliance. Instead, user space is provided by the host OS, the container image base layer or a sidecar container.
+
+### CLI
+* The `microshift` binary runs the Control Plane / Node process, it is not a tool to manage or be clients to those processes (like `oc` or `kubeadmin`). This is reflected in the sub-commands and paraemters offered by it, e.g. using the `run` verb (which implies run-to-cancel/run-to-completion) instead of `start`/`stop` verb-pairs (which imply asynch commands that return immediately).
+* For consistency and to play nicely with systemd, we avoid command line parameters that would need to be different between invokations (e.g. first-run vs subsequent runs) or instantiations (e.g. 1st Control Plane instances vs. 2nd or 3rd Control Plane instance).
+
+### Configuration
+* MicroShift uses a strictly declarative style of configuration.
+* MicroShift uses as few configuration options as possible. Where it provides configuration options, they are intuitive and have sensible defauls respectively are auto-configured.
+* MicroShift is preferably configured through config files, but allows overriding of parameters via environment variables (for use in containers, systemd) and command line flags (for ad-hoc use).
+* MicroShift can use both user-local and system-wide configuration.
+
+### Security
+* MicroShift instances should run with least privileges. In particular Control Plane-only instances should run completely non-privileged.
+* MicroShift should minimize the number of open network ports. In particular Node-only instances should not open any listening ports.
+* Assume there is no way to access the device or MicroShift instance from remote, i.e. no SSH access nor Kubernetes API access (for kubectl). Instead, communication is *always* initiated from the device / cluster towards the management system.
+* Open issues / questions:
+  * Model for joining nodes to MicroShift clusters.
+  * Model for certificate rotation.
+  * Requirements for FedRAMP and DISA STIG compliance.
+  * How to leverage / support the host OS's features for FIDO Device Onboard, remote attestation, integrity measurement architecture.
+  * How to support secure supply chain.
+
+### Networking
+* Host networking is configured by device management. MicroShift has to work with what it's been given by the host OS.
+* No Multus.
+* Open issues / questions:
+  * Lightweight CNI?
+  * API Load Balancing?
+  * Service Load Balancing?
+  * Ingress?
+
+### Storage
+* MicroShift defaults to local ephemeral storage (enough for basic use cases).
+* Open issues / questions:
+  * Provide escape hatch to add own CSI (which?).
```

Comment:
I believe something like NFS is internally provided by k8s, but I will try to spin it with our current release of microshift.

Reply:
https://github.com/kubernetes/examples/blob/master/staging/volumes/nfs/nfs-pv.yaml
https://github.com/kubernetes/examples/blob/master/staging/volumes/nfs/nfs-pvc.yaml
(at least via manual provisioning)

Reply:
I would need to check this works on microshift today without changes, then I can submit a follow up PR to comment on this may be?

Reply:
We don't need to solve all the questions now. This is just an initial PR to document the design decisions we've made earlier and open questions we are already aware of. I'd expect this to be a living document that we add to over time as our understanding increases.